### PR TITLE
cli/tests: Prevent git.subprocess from reading outside git config

### LINKED ISSUE #6159: Tests failing: Snapshot tests using git
## Steps to Reproduce the Problem

1. `git clone git@github.com:jj-vcs/jj.git`
1. `cd jj`
1. `cargo insta test --workspace`

## Expected Behavior

Tests pass.

## Actual Behavior

Getting lots of test failures (the ones I saw seem to be related to running `git`?):

```
error: 1 target failed:
    `-p jj-cli --test runner`
info: 36 snapshots to review
use `cargo insta review` to review snapshots
```

```
Reviewing [1/36] jj-cli@0.27.0:
Snapshot: git_clone
Source: cli/tests/test_git_clone.rs:57
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Expression: output
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
    0     0 │ ------- stderr -------
    1       │-Fetching into new repo in "$TEST_ENV/empty"
    2       │-Nothing changed.
    3       │-[EOF]
          1 │+Error: Git remote named 'origin' already exists
          2 │+[EOF]
          3 │+[exit status: 1]
────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

  a accept     keep the new snapshot
  r reject     retain the old snapshot
  s skip       keep both for now
  i hide info  toggles extended snapshot info
  d hide diff  toggle snapshot diff
```

Things I tried:
- Use a fresh clone/clean target folder
- Run tests from bash instead of fish

## Specifications

- MacOS 15.3.2 (24D81)
- git 2.49.0
- cargo 1.84.1 (66221abde 2024-11-19)
- rustc 1.84.1 (e71f9a9a9 2025-01-27)
- cargo-insta 1.42.2

### PR DESCRIPTION
**Pull request by robinst**
A subset of cli tests could fail if the system `/etc/gitconfig` had configuration interfering with the tests. The cause seems to be running of `jj` commands that would in turn use a `git` subprocess.

Fix this by setting `GIT_CONFIG_SYSTEM` and `GIT_CONFIG_GLOBAL`, like in `hermetic_git`.

Fixes #6159

# Checklist

If applicable:
- ~I have updated `CHANGELOG.md`~
- ~I have updated the documentation (README.md, docs/, demos/)~
- ~I have updated the config schema (cli/src/config-schema.json)~
- ~I have added tests to cover my changes~

### CODE REVIEW

**Review by yuja — APPROVED** [review]

**Comment by robinst** [comment]
Getting a failing test, but just on [test (windows-x86_64)](https://github.com/jj-vcs/jj/actions/runs/14142948254/job/39630587144?pr=6171#logs):

```
──── STDOUT:             jj-cli::runner test_git_remotes::test_git_remote_with_branch_config
running 1 test
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: git_remote_with_branch_config-3
Source: D:\a\jj\jj:450
───────────────────────────────────────────────────────────────────────────────
Expression: read_git_config(work_dir.root())
───────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬──────────────────────────────────────────────────────────────────
    3     3 │ 	logallrefupdates = false
    4     4 │ [branch "test"]
    5     5 │ 	remote = bar
    6     6 │ 	merge = refs/heads/test
          7 │+
          8 │+
          9 │+
    7    10 │ [remote "bar"]
    8    11 │ 	url = http://example.com/repo
    9    12 │ 	fetch = +refs/heads/*:refs/remotes/bar/*
────────────┴──────────────────────────────────────────────────────────────────
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
test test_git_remotes::test_git_remote_with_branch_config ... FAILED
failures:
failures:
    test_git_remotes::test_git_remote_with_branch_config
test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 820 filtered out; finished in 5.46s
```

Retried once and got the same failure, so doesn't look like a flake. Any ideas?

**Comment by yuja** [comment]
> ```
>           7 │+
>           8 │+
>           9 │+
> ```

Gitoxide might process non-existing path differently. `/dev/null` isn't a valid file on Windows.

Maybe we can filter out blank lines in `read_git_config()`?

File: cli/tests/common/test_environment.rs [hunk 1]
```diff
@@ -120,6 +120,14 @@ impl TestEnvironment {
         // executables like `git` from the PATH.
         cmd.env("PATH", std::env::var_os("PATH").unwrap_or_default());
         cmd.env("HOME", self.home_dir.to_str().unwrap());
+        // Prevent git.subprocess from reading outside git config
+        let null_device = if cfg!(windows) {
+            "NUL"
+        } else {
+            "/dev/null"
+        };
```
Comment (martinvonz) [inline cli/tests/common/test_environment.rs:128]: Does it work to use an empty string on all platforms?

Comment (robinst) [reply cli/tests/common/test_environment.rs]: No, same problem.

I wonder if it has something to do with `\r\n` vs `\n`. `test_git_remote_with_branch_config` adds to the config file with `writeln!`, which [always just uses LF](https://doc.rust-lang.org/std/macro.writeln.html). If the beginning of the file uses `\r\n` but then later switches to `\n`, maybe that confuses Gitoxide? But then why would setting the env variables affect that 🤔.

Let me try and see if writing `\r\n` fixes it.

Comment (robinst) [reply cli/tests/common/test_environment.rs]: Huh, that worked (even on non-Windows platforms). It feels like a bug/edge case in Gitoxide? I would report it but I don't have a Windows box to produce a minimal example.

Comment (yuja) [reply cli/tests/common/test_environment.rs]: Our test might depend on the system git config (which would probably use LF on Windows), and gix might choose the platform-native line ending if there was no existing config files to derive style from. And gix might be strict about line ending chars when reformatting the config file.

Anyway, I think we can filter out uninteresting lines in tests.

**Review by martinvonz — APPROVED** [review]
